### DISCUSSION TITLE
Task - MongoDB - Operators - Comparision Query Operators

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#27] [Tasks]
You are given the collection of `products` with the following documents:
```
db.products.insertMany([
    { "_id" : 1, "name" : "xPhone", "price" : 799, "releaseDate": ISODate("2011-05-14"), "spec" : { "ram" : 4, "screen" : 6.5, "cpu" : 2.66 },"color":["white","black"],"storage":[64,128,256]},
    { "_id" : 2, "name" : "xTablet", "price" : 899, "releaseDate": ISODate("2011-09-01") , "spec" : { "ram" : 16, "screen" : 9.5, "cpu" : 3.66 },"color":["white","black","purple"],"storage":[128,256,512]},
    { "_id" : 3, "name" : "SmartTablet", "price" : 899, "releaseDate": ISODate("2015-01-14"), "spec" : { "ram" : 12, "screen" : 9.7, "cpu" : 3.66 },"color":["blue"],"storage":[16,64,128]},
    { "_id" : 4, "name" : "SmartPad", "price" : 699, "releaseDate": ISODate("2020-05-14"),"spec" : { "ram" : 8, "screen" : 9.7, "cpu" : 1.66 },"color":["white","orange","gold","gray"],"storage":[128,256,1024]},
    { "_id" : 5, "name" : "SmartPhone", "price" : 599,"releaseDate": ISODate("2022-09-14"), "spec" : { "ram" : 4, "screen" : 9.7, "cpu" : 1.66 },"color":["white","orange","gold","gray"],"storage":[128,256]}
 ])
```
Perform and demonstrate the usage of the following comparison query operators:
1. $eq
2. $lt
3. $lte
4. $gt
5. $gte
6. $ne
7. $in
8. $nin

## Operations to be performed
1. Search for documents where the value of the ram field in the spec document equals 4
2. Query the products collection to find all documents where the array color contains an element with the value "black"
3. Select documents where the value of the screen field in the spec document is less than 7
4. Query the products collection to find all documents where the array storage has at least one element less than 128
5. Select documents where the value of the screen field in the spec document is less than or equal to 6.5
6. Select documents from the products collection to find all documents where the release date is before or on 2015-01-11
7. Select documents where the value of the screen field in the spec document is not equal to 9.7
8. Find documents where the color array has at least one element that matches either /^g+/ or /^w+/ regular expression
9. Select documents where the color array doesn’t have an element that is either "black" or "white"

**Comment by skagarwal5555** [comment]
```
1. 
db.products.find({"spec.ram" : 4}).pretty();
db.products.find( { "spec.ram" : { "$eq": 4}}).pretty();

2.
db.products.find({ "color" : { "$in" : ["black"]}}).pretty();

3.
db.products.find({ "spec.screen" : { "$lt" : 7}}).pretty();

4.
db.products.find({ "storage" : { "$lt" : 128 }}).pretty();

5.
db.products.find({ "spec.screen" : { "$lte" : 6.5}}).pretty();

6.
db.products.find({ "releaseDate" : { "$lte" : new Date("2015-01-11")}}).pretty();

7.
db.products.find({ "spec.screen" : { "$ne" : 9.7}}).pretty();

8.
db.products.find({ $or : [ {"color" : /^g+/}, { color : /^w+/}]} ).pretty();

9.
db.products.find({ "color" : { "$nin" : ["black","white"]}}).pretty();
```

**Comment by Praki87m** [comment]
1. db.products.find({"spec.ram": { $eq: 4}}).pretty()
2. db.products.find({color: { $eq: "black"}}).pretty()
3. db.products.find({"spec.screen": { $lt: 7}}).pretty()
4. db.products.find({storage: { $lt: 128}}).pretty()
5. db.products.find({"spec.screen": { $lte: 6.5}}).pretty()
6. db.products.find({releaseDate: { $lte: new Date('2015-01-11')}}).pretty()
7. db.products.find({screen: { $ne: 9.7}}).pretty()
8. db.products.find({"color": {$regex: /^g/, $regex: /^w/}}).pretty()
9. db.products.find({color: { $ne: "black",$ne: "white" }}).pretty()

**Comment by sanath8se** [comment]
db.products.find({"spec.ram": {$eq: 4}});
db.products.find({color: {$eq: "black"}});
db.products.find({"spec.screen": {$lt: 7}});
db.products.find({storage: {$lt: 128}});
db.products.find({"spec.screen": {$lte: 6.5}});
db.products.find({releaseDate: {$lte: ISODate("2015-01-11")}});
db.products.find({"spec.screen": {$ne: 9.7}});
db.products.find({color: {$in: [/^g+/, /^w+/]}});
db.products.find({color: {$nin: ["black", "white"]}});

**Comment by uday2305** [comment]
1. db.products.find({"spec.ram":{$eq:4}}).pretty()
2. db.products.find({color:{$in:["black"]}}).pretty()
3. db.products.find({"spec.screen":{$lt:7}}).pretty()
4. db.products.find({"storage":{$lt:128}}).pretty()
5. db.products.find({"spec.screen":{$lte:6.5}}).pretty()
6. db.products.find({"releaseDate":{$lt:ISODate("2015-01-11")}}).pretty()
7. db.products.find({"spec.screen":{$ne:9.7}}).pretty()
8. db.products.find({color:{$in:[/^g+/,/^w+/]}}).pretty()
9. db.products.find({color:{$nin:["black","white"]}}).pretty()

**Comment by Manogaran-Coded** [comment]
> db.products.find( { "spec.ram": { $eq: 4 } } )
{ "_id" : 1, "name" : "xPhone", "price" : 799, "releaseDate" : ISODate("2011-05-14T00:00:00Z"), "spec" : { "ram" : 4, "screen" : 6.5, "cpu" : 2.66 }, "color" : [ "white", "black" ], "storage" : [ 64, 128, 256 ] }
{ "_id" : 5, "name" : "SmartPhone", "price" : 599, "releaseDate" : ISODate("2022-09-14T00:00:00Z"), "spec" : { "ram" : 4, "screen" : 9.7, "cpu" : 1.66 }, "color" : [ "white", "orange", "gold", "gray" ], "storage" : [ 128, 256 ] }
> db.products.find( { "spec.screen": { $lt: 4 } } )
> db.products.find( { "spec.screen": { $lt: 7 } } )
{ "_id" : 1, "name" : "xPhone", "price" : 799, "releaseDate" : ISODate("2011-05-14T00:00:00Z"), "spec" : { "ram" : 4, "screen" : 6.5, "cpu" : 2.66 }, "color" : [ "white", "black" ], "storage" : [ 64, 128, 256 ] }
> db.products.find( { "spec.screen": { $lte: 7 } } )
{ "_id" : 1, "name" : "xPhone", "price" : 799, "releaseDate" : ISODate("2011-05-14T00:00:00Z"), "spec" : { "ram" : 4, "screen" : 6.5, "cpu" : 2.66 }, "color" : [ "white", "black" ], "storage" : [ 64, 128, 256 ] }``

**Comment by MRamitkv** [comment]
> db.products.find({"spec.ram": 4})
> db.products.find({color:"black"})
> db.products.find({"spec.screen":{"$lt":7}})
> db.products.find({storage:{"$lt":128}})
> db.products.find({"spec.screen":{"$lte":6.5}} )
> db.products.find({releaseDate:{"$lte": ISODate("2015-01-11")}})
> db.products.find({"spec.screen":{"$ne":9.7}})
> db.products.find({color:{$in:[/^w+/, /^g+/]}})
> db.products.find({color:{"$nin":["black", "white"]}})
``

**Comment by shubhamk-coder** [comment]
@akash-coded 
1. $eq : Specifies equality condition. The $eq operator matches documents where the value of a field equals the specified value.
         Syntax : { <field>: { $eq: <value> } } 

   -> db.products.find({"spec.ram" : { $eq : 4 }})

2. $in : The $in operator selects the documents where the value of a field equals any value in the specified array. To specify an $in expression, use the following prototype:
         Syntax : { field: { $in: [<value1>, <value2>, ... <valueN> ] } }

   -> db.products.find({color : { $in : "black" }})

3. $lt : $lt selects the documents where the value of the field is less than (i.e. <) the specified value.
	 Syntax : { field: { $lt: value } } 

   -> db.products.find({"spec.screen" : { $lt : 7 }})

4. $lt : $lt selects the documents where the value of the field is less than (i.e. <) the specified value.
	 Syntax : { field: { $lt: value } } 

   -> db.products.find({storage : { $lt : 128 }})

5. $lte : $lte selects the documents where the value of the field is less than or equal to (i.e. <=) the specified value.
	 Syntax : { field: { $lte: value } } 

   -> db.products.find({"spec.screen" : { $lte : 6.5 }})

6. $lte : $lte selects the documents where the value of the field is less than or equal to (i.e. <=) the specified value.
	 Syntax : { field: { $lte: value } } 

   -> db.products.find({releaseDate: { $lte : new Date('2015-01-11')}})

7. $ne : $ne selects the documents where the value of the field is not equal to the specified value. This includes documents that do not contain the field.
	 Syntax : { field: { $ne: value } }

   -> db.products.find({"spec.screen" : { $ne : 9.7 }})

8. $eq : Specifies equality condition. The $eq operator matches documents where the value of a field equals the specified value.
         Syntax : { <field>: { $eq: <value> } }
	 If <value> is a regular expression, the statement is expanded using the $regex operator { field: { $regex: <value> } }.

   -> db.products.find({color : { $regex : [ /^g+/, /^w+/ ] }})

9. $nin : $nin selects the documents where:
		- the field value is not in the specified array or
		- the field does not exist.
	  If the field holds an array, then the $nin operator selects the documents whose field holds an array with no element equal to a value in the specified array (for example, <value1>, <value2>, and so on).
	  Syntax: { field: { $nin: [ <value1>, <value2> ... <valueN> ] } }

   -> db.products.find({color : { $nin : [ "black", "white" ] }})

Remaining :-

10. $gt : $gt selects the documents where the value of the field is greater than (i.e. >) the specified value.
	 Syntax : { field: { $gt: value } } 

11. $gte : $gte selects the documents where the value of the field is greater than or equal to (i.e. >=) the specified value.
	 Syntax : { field: { $gte: value } }

**Comment by Thharshita** [comment]
1. Search for documents where the value of the ram field in the spec document equals 4
>db.products.find({
"spec.ram":{
$eq:4
}
})
2. Query the products collection to find all documents where the array color contains an element with the value "black"
db.products.find({
"color":{
$eq:["black"]
}
})
3. Select documents where the value of the screen field in the spec document is less than 7
>db.products.find{(
"spec.screen":{
$lt:7
}
})
4. Query the products collection to find all documents where the array storage has at least one element less than 128
>db.products.find({
"storage":{
$lt:128
}
})
5. Select documents where the value of the screen field in the spec document is less than or equal to 6.5
>db.products.find({
"spec.screen":{
$lte:6.5
}
})
6. Select documents from the products collection to find all documents where the release date is before or on 2015-01-11
>db.products.find({
"releaseDate":{
$lte:new Date('2015-01-11')
}
})
7. Select documents where the value of the screen field in the spec document is not equal to 9.7
>db.products.find({
"spec.screen":{
$ne:9.7
}
})
8. Find documents where the color array has at least one element that matches either /^g+/ or /^w+/ regular expression
>db.products.find({
"color":{
$in:[ /^g+/ ,/^w+/ ]
}
})

>db.products.find({
"color":{
$regex:[ /^g+/ ,/^w+/ ]
}
})
9. Select documents where the color array doesn’t have an element that is either "black" or "white"
>db.products.find({
"color":{
$rnin:["black","white"]
}
})
